docs/releases/3.2.txt

```diff
@@ -52,6 +52,16 @@ With automatic ``AppConfig`` discovery, ``default_app_config`` is no longer
 needed. As a consequence, it's deprecated.
 
 See :ref:`configuring-applications-ref` for full details.
+Expressions alias
+------------------
+
+Expressions that don't need to be selected can now be saved as an alias for
+later reuse with other ``QuerySet`` methods via a new
+:meth:`~django.db.models.query.QuerySet.alias` method. This is useful when
+the result of the expression itself is not needed but is used for filtering,
+ordering, or as a part of a complex expression. By not selecting the actual value, one can construct
+a smaller SQL query, return less data from the database, and avoid double evaluation
+of costly expressions, which results in better performance.
```

Comment:
Much of the motivation for me and the original issue author was the query execution time, and query construction time by the ORM. When using `alias()` with `filter()`, the expression won't be included both in the SELECT and WHERE clauses as with `annotate()`. Also, it is not rare to make some annotations that may or may not be used for filtering and ordering later, but most of them are discarded, so making them `alias` will improve performance.
I am not sure whether any backend will detect duplicated expressions and cache the results, but it seems that avoiding duplicated expressions leads to a significant performance boost, especially when annotations are `Subquery`.
So, I believe that performance boost due to the fewer data returned is not the most important one, and we should emphasize on the query execution time.

> Doing a full sequential scan (no index) with both calculating the to_tsvector() for return and in query predicate is 27,000ms without recalculating the to_tsvector for return, 46,700ms with recalculating it.
> After having created a functional index on the to_tsvector() call, and using a relatively common search term, the time to return without recalculating the to_tsvector was 1.7ms, vs 21,200ms with recalculating the ts_vector.

Reply:
> query construction time by the ORM.

I'm not sure this is actually significant, we're talking about a single `as_sql` call removal here since the expression will be resolved once in both cases. I mean, sure that's one less expression compilation down the line but I don't think it's significant enough to warrant a mention in the docs.

Speaking of which we should probably add a test to make sure that `JOIN` added for an unused alias is pruned 

e.g. `Book.objects.alias(author_name=Lower('author__name'))` should not create a `JOIN`.

> Also, it is not rare to make some annotations that may or may not be used for filtering and ordering later, but most of them are discarded, so making them alias will improve performance.

I agree that it isn't rare but I wouldn't say it's common either. From experience I can count the number of times this has come up on the past projects I worked on a single hand.

> I am not sure whether any backend will detect duplicated expressions and cache the results, but it seems that avoiding duplicated expressions leads to a significant performance boost, especially when annotations are Subquery.

While this is mentioned on the ticket this is now easy to work around by passing `Subquery` and other such expressions directly to `filter` or `order_by` which should cover 90% of the cases thanks to ticket-25367 being fixed.

> So, I believe that performance boost due to the fewer data returned is not the most important one, and we should emphasize on the query execution time.

I agree that the above statement should be adapted to reflect this. What about

> Not selecting the actual value can allow the database to perform some optimization and result in less data being returned. This can result in better performance when dealing with certain type of columns or complex expressions.

Reply:
I think I will need to come up with some performance tests for query construction to make a final decision, but I've experienced significant CPU time cost due to the query size. 

> Speaking of which we should probably add a test to make sure that JOIN added for an unused alias is pruned

As stated in the ticket, 
> all other logic remains the same: despite the expression is not selected, the joins and group by are the same as for annotation. 

However, it seems like LEFT OUTER joins with FK can be pruned, I'll take a look and the current behavior.

About the next two points: when the complex expression is constructed by parts, in different places and only later somewhere may be used for filtering or sorting, most of the annotations are useless, because only final expression is valuable, and that expression can be used only in filter clause, not required to select. That's basically the whole point of the ticket, to produce `SELECT x, y FROM z WHERE (SubExpression1 + SubExpression2 + SubExpression3)  = FOO` instead of 
`SELECT x, y, SubExpression1 as a, SubExpression2 as b, SubExpression3 as c, (SubExpression1 + SubExpression2 + SubExpression3) as d  FROM z WHERE (SubExpression1 + SubExpression2 + SubExpression3) = FOO`
where each SubExpression may be a `Subquery` or another costly expression.

For me, the ticket-25367 doesn't cover most of the usages, because often you make annotations and filtering in completely different parts of the codebase, i.e: annotate in `QuerySet`, filter in the view.

I think it is hard to explain most use cases and benefits, we may leave that to users and blog posts authors. But I believe that the query without expressions in the SELECT will perform better, refer to the ticket for some examples of 1.7ms vs 21200ms due to SELECT. What about

> Not selecting the unused value removes redundant work from the database which leads to better performance.

Reply:
> That's basically the whole point of the ticket ...
> For me, the ticket-25367 doesn't cover most of the usages, because often you make annotations and filtering in completely different parts of the codebase, i.e: annotate in QuerySet, filter in the view.

I would argue that's bad practice to add unused annotations at the queryset/CBV level if it's only going to be used for filtering or ordering purpose under certain circumstances. I suspect folks who are investing time identifying the origin of some slow queries generated by the ORM are already avoiding this pattern.

> What about
>> Not selecting the unused value removes redundant work from the database which should result in better performance.

That works for me.

Reply:
About the JOIN - it is not trimmed, and I'm very unsure it can be done without significant effort. From the beginning, I've decided this is actually an intended behavior, so I suggest to keep it as is unless someone can propose a simple solution without caveats.

Reply:
> From the beginning, I've decided this is actually an intended behavior, so I suggest to keep it as is unless someone can propose a simple solution without caveats.

Makes sense.